### PR TITLE
Fix and clean up persistent login.

### PR DESCRIPTION
**Pull request by EreMaijala**
- Adds support for more lenient token rotation to fix cookie persistence issues with Safari.
- Adds logging.
- Removes an invalid token cookie.
- Fixes deletion of a series if not currently logged in a persistent session.
- Changes scope of some public methods to protected.

Safari appears to have an issue (tested with macOS 14.4 and iOS 17.3.1) where it doesn't always persist cookies or updates to their values properly. That's why this introduces a bit more lenient approach that allows token re-use until the next one is received. The new approach is enabled by default but can be disabled as required.

### CODE REVIEW

File: module/VuFind/src/VuFind/Db/Table/LoginToken.php [hunk 1]
```diff
@@ -109,31 +110,42 @@ public function saveToken(
      */
     public function matchToken(array $token): ?LoginTokenRow
     {
-        $row = $this->getBySeries($token['series'], $token['user_id']);
-        if ($row && hash_equals($row['token'], hash('sha256', $token['token']))) {
-            if (time() > $row['expires']) {
-                $row->delete();
-                return null;
+        $rowsFound = false;
+        foreach ($this->select(['user_id' => $token['user_id'], 'series' => $token['series']]) as $row) {
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Db/Table/LoginToken.php]: I see that you factored this to stop using the getBySeries method (and I understand why). Does this mean that getBySeries is now useless and should be removed? Or should we keep the method and add an argument to switch between returning "current" and "all" so we can still use it here?

File: module/VuFind/src/VuFind/Auth/LoginTokenManager.php [hunk 2]
```diff
@@ -251,18 +266,22 @@ public function requestIsFinished(): void
     /**
      * Create a new login token
      *
-     * @param \VuFind\Db\Row\User $user      user
-     * @param string              $series    login token series
-     * @param string              $sessionId Session identifier
-     * @param int                 $expires   Token expiration timestamp
+     * @param User   $user           User
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Auth/LoginTokenManager.php]: Is it worth mentioning in the comment that this also has the side effect of deleting outdated tokens? Now that the functionality has been combined, the method name might be misleading.

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks for the progress on this, @EreMaijala. See below for a few comments. Also:

1.) As mentioned in Slack, I had trouble with the console browscap tool failing to update things to a working state; the Admin module piece worked as expected.

2.) The profile page fails when using PostgreSQL as the database backend:

```
2024-03-12T12:42:28-04:00 CRIT (2): Laminas\Db\Adapter\Exception\InvalidQueryException : Statement could not be executed (42803 - 7 - ERROR:  column "login_token.browser" must appear in the GROUP BY clause or be used in an aggregate function

LINE 1: ... AS "user_id", MAX("last_login") AS "last_login", "login_tok...

                                                             ^) ; PDOException : SQLSTATE[42803]: Grouping error: 7 ERROR:  column "login_token.browser" must appear in the GROUP BY clause or be used in an aggregate function

LINE 1: ... AS "user_id", MAX("last_login") AS "last_login", "login_tok...
```

**Comment by demiankatz** [comment]
Update: the reason the command line tool didn't work for me is that it created all the files with 600 permissions, and then Apache couldn't read them. Any idea why that might be? I'd think we at least want 644...

File: module/VuFind/src/VuFind/Db/Table/LoginToken.php (hunk 1 above)
Comment (EreMaijala) [reply module/VuFind/src/VuFind/Db/Table/LoginToken.php]: Now getBySeries returns all in the series and is again used here as well.

File: module/VuFind/src/VuFind/Auth/LoginTokenManager.php (hunk 2 above)
Comment (EreMaijala) [reply module/VuFind/src/VuFind/Auth/LoginTokenManager.php]: Renamed, made protected and added a simplified public createToken.

**Comment by EreMaijala** [comment]
> Update: the reason the command line tool didn't work for me is that it created all the files with 600 permissions, and then Apache couldn't read them. Any idea why that might be? I'd think we at least want 644...

We have the permissions to use in config.ini. I suppose it would be possible to override them for browscap cache, but I dared not do it.

File: module/VuFind/src/VuFind/Db/Table/LoginToken.php [hunk 3]
```diff
@@ -168,13 +202,12 @@ public function getByUserId(int $userId): array
...
-     * @param int    $userId User identifier
      *
-     * @return ?LoginTokenRow
+     * @return ResultSetInterface
      */
-    public function getBySeries(string $series, int $userId): ?LoginTokenRow
+    public function getBySeries(string $series): ResultSetInterface
     {
-        return $this->select(['user_id' => $userId, 'series' => $series])->current();
+        return $this->select(['series' => $series]);
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Db/Table/LoginToken.php]: This could be a compact now.

File: module/VuFind/src/VuFind/Auth/LoginTokenManager.php [hunk 4]
```diff
@@ -212,6 +227,20 @@ public function tokenLogin(string $sessionId): ?\VuFind\Db\Row\User
         return $user;
     }
 
+    /**
+     * Rotate login token in given series or create a new series
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Auth/LoginTokenManager.php]: Is this comment accurate for this method? If this is also rotating, does it make sense to have a separate method name for the public version?

File: module/VuFind/src/VuFind/Auth/LoginTokenManager.php [hunk 5]
```diff
@@ -336,20 +345,74 @@ public function deleteUserLoginTokens($userId)
...
         $this->cookieManager->clear($this->getCookieName());
     }
 
+    /**
+     * Create a new login token series or rotate login token in given series
+     *
+     * @param User   $user           User
+     * @param string $sessionId      Session identifier
+     * @param string $series         Login token series
+     * @param int    $expires        Token expiration timestamp
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/Auth/LoginTokenManager.php]: Should this comment note the special-case behavior for a 0 value?

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @EreMaijala! See below for a few minor comments (mostly documentation-related). Also note that the Profile page is still broken in PostgreSQL due to what appears to be a problem with SQL grouping behavior (exact error message in my previous review comment).

File: module/VuFind/src/VuFind/Db/Table/LoginToken.php (hunk 3 above)
Comment (EreMaijala) [reply module/VuFind/src/VuFind/Db/Table/LoginToken.php]: Fixed.

File: module/VuFind/src/VuFind/Auth/LoginTokenManager.php (hunk 4 above)
Comment (EreMaijala) [reply module/VuFind/src/VuFind/Auth/LoginTokenManager.php]: Oops. Revised.

File: module/VuFind/src/VuFind/Auth/LoginTokenManager.php (hunk 5 above)
Comment (EreMaijala) [reply module/VuFind/src/VuFind/Auth/LoginTokenManager.php]: I changed the special case to null and added a note.

**Review by demiankatz — APPROVED** [review]
Thanks, @EreMaijala, this looks good to me now, and everything is working in my test environment with PostgreSQL.

**Pull request merged**